ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
     if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
       List<String> potentialLogSegments = blobStore.getLogSegmentsNotInJournal();
       if (potentialLogSegments != null) {
-        details = new CompactionDetails(time.milliseconds() - messageRetentionTimeInMs, potentialLogSegments);
+        // fetch valid data size from BlobStoreStats
+        BlobStoreStats blobStoreStats = blobStore.getBlobStoreStats();
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
+
+        // remove those log segments having overlapping entries in journal
+        List<String> invalidLogSegments = new ArrayList<>();
+        for (String segmentName : validDataPerLogSegments.getSecond().keySet()) {
+          if (!potentialLogSegments.contains(segmentName)) {
+            invalidLogSegments.add(segmentName);
+          }
+        }
+        for (String invalidLogSegment : invalidLogSegments) {
```

Comment:
if you use an iterator, you could remove in a single loop. There will be no need to have the temporary `invalidLogSegments` variable.